cirq-core/cirq/protocols/apply_unitary_protocol.py

```diff
@@ -363,6 +374,7 @@ def apply_unitary(
     with warnings.catch_warnings():
         warnings.filterwarnings(action="error", category=np.ComplexWarning)
         for strat in strats:
+            strat = cast(Callable[[Any, ApplyUnitaryArgs], Optional[np.ndarray]], strat)
```

Comment:
I don't think signature is true for all the strategies? `_strat_apply_unitary_from_apply_unitary` and `_strat_apply_unitary_from_decompose` both don't take the optional matrix parameters. 

Maybe create a helper method for `_strat_apply_unitary_from_unitary` and dispatch to it with the matrix parameter wherever needed?

Reply:
done